### PR TITLE
Change to use buffer when writing binary/byte[]

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

* Improve the json writer process for the binary/byte[].

### Description

* Improve the json writer process for the binary/byte[].

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Comment by xuzhg** [comment]
@ysanghi Would you please take a look?

File: src/Microsoft.OData.Core/Json/JsonValueUtils.cs [hunk 1]
```diff
@@ -327,11 +328,43 @@ internal static void WriteValue(TextWriter writer, byte[] value)
...
+            for (int i = 0; i < value.Length; i += bufferByteSize)
+            {
+                int length = bufferByteSize;
+                if (i + length > value.Length)
+                {
+                    length = value.Length - i;
+                }
+
+                int output = Convert.ToBase64CharArray(value, i, length, buffer, 0);
+                writer.Write(new string(buffer, 0, output));
```
Comment (ysanghi) [inline src/Microsoft.OData.Core/Json/JsonValueUtils.cs]: Why create a string here. Can we avoid this?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonValueUtilsTests.cs [hunk 2]
```diff
@@ -241,24 +241,38 @@ public void WriteByteShouldWork()
         {
             this.TestInit();
             var byteArray = new byte[] { 1, 2, 3 };
-            JsonValueUtils.WriteValue(this.writer, byteArray);
+            JsonValueUtils.WriteValue(this.writer, byteArray, ref this.buffer);
             this.StreamToString().Should().Be("\"AQID\"");
         }
 
+        [Fact]
+        public void WriteLongBytesShouldWork()
```
Comment (ysanghi) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonValueUtilsTests.cs]: Write edge case unit tests where size of the buffer is equal to the length of the string, etc..

File: src/Microsoft.OData.Core/Json/JsonValueUtils.cs (hunk 1 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/Json/JsonValueUtils.cs]: Yes. We can. I update to call `writer.Write(buffer, 0, output);` #Resolved.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonValueUtilsTests.cs (hunk 2 above)
Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonValueUtilsTests.cs]: # Added.

File: src/Microsoft.OData.Core/Json/JsonValueUtils.cs [hunk 3]
```diff
@@ -327,11 +328,43 @@ internal static void WriteValue(TextWriter writer, byte[] value)
...
+
+            buffer = BufferUtils.InitializeBufferIfRequired(buffer);
+            Debug.Assert(buffer != null);
+
+            int bufferLength = buffer.Length;
+
+            // Try to hold base64 string as much as possible in one converting.
+            int bufferByteSize = bufferLength * 3 / 4;
+
+            for (int i = 0; i < value.Length; i += bufferByteSize)
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/Json/JsonValueUtils.cs]: >i [](start = 21, length = 1)

Consider using a more meaningful name, like "startPos"

Comment (xuzhg) [reply src/Microsoft.OData.Core/Json/JsonValueUtils.cs]: Thanks, I change to use "offsetIn", which is used in 'ToBase64CharArray`

Comment (mikepizzo) [reply src/Microsoft.OData.Core/Json/JsonValueUtils.cs]: perfect
